File: dsa-problems/Strings/solutions/java/STR-012-distinct-subsequence-char-limit.java
class Solution {
    public int countDistinctSubsequencesWithLimit(String s, int maxFreq, int MOD) {
        Map<String, Long> dp = new HashMap<>();
        dp.put(encode(new int[26]), 1L);

        for (char c : s.toCharArray()) {
            int charIdx = c - 'a';
            Map<String, Long> newDp = new HashMap<>();

            for (Map.Entry<String, Long> entry : dp.entrySet()) {
                String stateStr = entry.getKey();
                long count = entry.getValue();
                int[] state = decode(stateStr);

                // Don't include
                String key = stateStr;
                newDp.put(key, (newDp.getOrDefault(key, 0L) + count) % MOD);

                // Include if allowed
                if (state[charIdx] < maxFreq) {
                    state[charIdx]++;
                    String newKey = encode(state);
                    newDp.put(newKey, (newDp.getOrDefault(newKey, 0L) + count) % MOD);
                    state[charIdx]--;  // Restore
                }
            }

            dp = newDp;
        }

        long total = 0;
        for (long count : dp.values()) {
            total = (total + count) % MOD;
        }
        return (int) total;
    }

    private String encode(int[] freq) {
        StringBuilder sb = new StringBuilder();
        for (int f : freq) {
            sb.append(f).append(",");
        }
        return sb.toString();
    }

    private int[] decode(String s) {
        String[] parts = s.split(",");
        int[] freq = new int[26];
        for (int i = 0; i < 26; i++) {
            freq[i] = Integer.parseInt(parts[i]);
        }
        return freq;
    }
}

File: dsa-problems/Strings/solutions/javascript/STR-012-distinct-subsequence-char-limit.js
function countDistinctSubsequencesWithLimit(s, maxFreq, MOD) {
  const encode = (freq) => freq.join(",");
  const decode = (str) => str.split(",").map(Number);

  let dp = new Map();
  dp.set(encode(new Array(26).fill(0)), 1);

  for (let char of s) {
    const charIdx = char.charCodeAt(0) - 97;
    const newDp = new Map();

    for (let [stateStr, count] of dp) {
      const state = decode(stateStr);

      // Don't include
      const key = stateStr;
      newDp.set(key, ((newDp.get(key) || 0) + count) % MOD);

      // Include if allowed
      if (state[charIdx] < maxFreq) {
        state[charIdx]++;
        const newKey = encode(state);
        newDp.set(newKey, ((newDp.get(newKey) || 0) + count) % MOD);
        state[charIdx]--; // Restore
      }
    }

    dp = newDp;
  }

  let total = 0;
  for (let count of dp.values()) {
    total = (total + count) % MOD;
  }
  return total;
}
